### PR TITLE
[Performance] Introduce header and Rails caching for APIv3

### PR DESCRIPTION
**Pull request by oliverguenther**
This commit introduces cached project-scoped work packages schema.
Their caches are expired on a key based on
- the Project ID
- the Type ID
- the allowed custom fields for the given project

This commit only caches schema, not form endpoints.

This is closely related to both 
https://community.openproject.com/work_packages/18404/activity
https://community.openproject.com/work_packages/13702/activity

### CODE REVIEW

**Comment by TeatroIO** [comment]
I've prepared a stage to preview changes. <a href='http://oliverguenther-feature-schema-caching.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/' target='_blank'>Open stage</a> or <a href='http://oliverguenther-feature-schema-caching.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com//teatro.log' target='_blank'>view logs</a>.

File: lib/api/v3/work_packages/schema/work_package_schema_representer.rb [hunk 1]
```diff
@@ -92,9 +92,6 @@ def initialize(schema, context)
 
           schema :spent_time,
                  type: 'Duration',
-                 show_if: -> (_) do
-                   current_user_allowed_to(:view_time_entries, context: represented.project)
-                 end,
```
Comment (oliverguenther) [inline lib/api/v3/work_packages/schema/work_package_schema_representer.rb:131]: I've removed this part since I argue we do not have to explicitly hide the schema link. If the user is unable to write it, it will be shown in the schema with form embedded. This will not contain any sensitive information, only that there is a `spentTime` attribute

**Comment by oliverguenther** [comment]
@machisuji Could you integrate this into your schema changes? We'll need to extend the cache key for that. 

Since the serialized attribute settings should automatically touch the type's `updated_at`, we can simply use that value for a cache key.

File: lib/api/v3/work_packages/schema/work_package_schemas_api.rb [hunk 2]
```diff
@@ -56,25 +60,30 @@ def raise404
...
               end
 
               get do
-                @representer
+                cache([cache_key(@project.id, @type.id), @type.updated_at, @custom_fields]) do
+                  schema = TypedWorkPackageSchema.new(project: @project, type: @type)
+                  self_link = api_v3_paths.work_package_schema(@project.id, @type.id)
+                  WorkPackageSchemaRepresenter.create(schema,
+                                                      self_link: self_link,
+                                                      current_user: nil)
```
Comment (machisuji) [inline lib/api/v3/work_packages/schema/work_package_schemas_api.rb:85]: Why did you change `current_user` to `nil` here? When trying this I get an error for any request to this endpoint.

```
NoMethodError (undefined method `allowed_to?' for nil:NilClass):
  lib/api/decorators/single.rb:119:in `current_user_allowed_to'
  lib/api/caching/helpers.rb:19:in `block in cache'
  lib/api/caching/helpers.rb:17:in `cache'
  lib/api/v3/work_packages/schema/work_package_schemas_api.rb:78:in `block (3 levels) in <class:WorkPackageSchemasAPI>'
```

Comment (machisuji) [inline lib/api/v3/work_packages/schema/work_package_schemas_api.rb:85]: Well, I suppose it doesn't make sense to use current_user if you want to cache this. True. Then the question is why that representer would need the current user. Am I missing some change perhaps? I just cherry-picked these two commits into my config to test this stuff. :D

Comment (machisuji) [inline lib/api/v3/work_packages/schema/work_package_schemas_api.rb:85]: Ah, it's [costs](https://github.com/finnlabs/openproject-costs/blob/dev/lib/open_project/costs/engine.rb#L231-L240) that breaks this.

Comment (oliverguenther) [inline lib/api/v3/work_packages/schema/work_package_schemas_api.rb:85]: Ah, good point. I would argue again that this is public information and does not need to be hidden behind a user check. All we know is that the project / work package allows to set costs. If the user is allowed to access the schema, that differentiation should not matter.

Thus I've introduced the `nil` here to make it clear that the schema representer may receive nil users for when showing the schema without its embedded form.

What do you think?

Comment (machisuji) [inline lib/api/v3/work_packages/schema/work_package_schemas_api.rb:85]: Yeah I think that makes perfect sense.

Comment (machisuji) [inline lib/api/v3/work_packages/schema/work_package_schemas_api.rb:85]: Would you mind creating a pull request for costs then that adapts the `show-if` accordingly?

Comment (oliverguenther) [inline lib/api/v3/work_packages/schema/work_package_schemas_api.rb:85]: Will do!

**Comment by machisuji** [comment]
:+1:
